text/0000-hidden-impls.md

```diff
@@ -0,0 +1,957 @@
+- Feature Name: `hidden_impls`
+- Start Date: 2018-08-24
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Allow a visibility modifier on a `trait` implementation such that:
+
++ it is public wrt. coherence and overlap
++ it is not usable in contexts where it is *hidden* (not visible)
+
+An example:
+
+```rust
+#[derive(Clone)]
+struct Foo(..);
+
+crate impl Copy for Foo {} // Can only be used in the current crate!
+```
+
+# Motivation
+[motivation]: #motivation
+
+First, let's assume that we have a trait `Property` defined somewhere.
+This trait, which doesn't have to be inside your crate, is also visible
+outside of it. It also happens that your crate has defined and exported
+a type called `Thing`. The specifics of `Property` and `Thing` are not
+interesting for our purposes, but they exist.
+
+You now want to, define an implementation of `Property` for `Thing`.
+But for some reason, you are not willing, or able, to expose this implementation
+outside your crate and make it part of your semantic versioning guarantees.
+
+An example of such a scenario is defining a new-type wrapper `TestRng` around
+`XorShiftRng` of the `rand` crate. We want to define an implementation of
+`SeedableRng` for `TestRng` to take advantage of the logic of `FromEntropy`
+which is automatically provided for us when `Rng: SeedableRng`.
+But we also want to avoid exposing `SeedableRng` as a guarantee to the
+users of our library (`proptest`) right now. The allure of getting the logic
+provided when defining `SeedableRng` could also have caused us to prematurely
+and mistakenly provide a guarantee we were not ready to provide.
+
+[proptest#80]: https://github.com/AltSysrq/proptest/issues/80
+[regex#457]: https://github.com/rust-lang/regex/issues/457
+
+A further example from `proptest` is when the author of this RFC
+[mistakenly][proptest#80] exposed `impl From<SizeRange> for SizeRange`
```

Comment:
typo: I think you mean `for Range<usize>` here

Reply:
I do! Thanks :)